cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs

```diff
@@ -111,10 +113,15 @@ decimal = do
 
 -- | Asset name parser.
 assetName :: Parser AssetName
-assetName =
-    toAssetName <$> some alphaNum
-  where
-    toAssetName = AssetName . Text.encodeUtf8 . Text.pack
+assetName = do
+  hexText <- some hexDigit
+  note "AssetName deserisalisation failed" $
+    deserialiseFromRawBytesHex AsAssetName $ BSC.pack hexText
+
+note :: MonadFail m => String -> Maybe a -> m a
```

Comment:
Why have you re-implemented the same function twice? DRY - Do not Repeat Yourself.

Reply:
each one has type incompatible with others

Reply:
Parser has no MonadFail

Reply:
Sorry, Aeson.Parser has MonadFail, but not MonadError, which is used by Cardano.Prelude.note.

And I can't figure out the right place for MonadFail version, so I repeated the code to save ourselves from bad module structure.

Can you suggest a place for this function?

Reply:
I don't follow. The `note` in `cardano-api/src/Cardano/Api/Value.hs` has the type signature `note :: MonadFail m => String -> Maybe a -> m a` which is identical to the `note` in `cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs`.

Why not put `note` in `Cardano.Api.Utils`?

Reply:
`Cardano.Api.Utils` is internal in `cardano-api`, I use `Cardano.CLI.Helpers`.

Reply:
Code moved to `Cardano.Api.Utils`